multisig-ism/src/metadata.sw

```diff
@@ -12,17 +14,54 @@ pub struct MultisigMetadata {
     validators: Vec<EvmAddress>,
 }
 
-// TODO: fix packing
-fn eth_hash(hash: b256) -> b256 {
-    return keccak256(("\x19Ethereum Signed Message:\n32", hash));
-}
+// TODO: properly import
+const B256_BYTE_COUNT: u64 = 32u64;
+const U32_BYTE_COUNT: u64 = 4u64;
+const U8_BYTE_COUNT: u64 = 1u64;
+const EVM_ADDRESS_BYTE_COUNT: u64 = 20u64;
 
 fn domain_hash(origin: u32, mailbox: b256) -> b256 {
-    return keccak256((origin, mailbox, "HYPERLANE"));
+    let suffix = "HYPERLANE";
+    let suffix_len = 9;
+
+    let mut bytes = Bytes::with_length(U32_BYTE_COUNT + B256_BYTE_COUNT + suffix_len);
+
+    bytes.write_u32(0, origin);
+    bytes.write_b256(U32_BYTE_COUNT, mailbox);
+    bytes.write_packed_bytes(U32_BYTE_COUNT + B256_BYTE_COUNT, __addr_of(suffix), suffix_len);
+
+    bytes.keccak256()
 }
 
 pub fn commitment(threshold: u8, validators: Vec<EvmAddress>) -> b256 {
-    return keccak256((threshold, validators));
+    let num_validators = validators.len();
+
+    let mut bytes = Bytes::with_length(U8_BYTE_COUNT + num_validators * EVM_ADDRESS_BYTE_COUNT);
+    bytes.write_u8(0, threshold);
+
+    let mut index = 0;
+    while index < num_validators {
+        bytes.write_packed_bytes(
+            U8_BYTE_COUNT + index * EVM_ADDRESS_BYTE_COUNT,
+            __addr_of(validators.get(index).unwrap()),
```

Comment:
Hmm this is interesting - I believe the EvmAddress type under the hood is just a b256, where the rightmost 20 bytes are the EVM address, and the leftmost 12 bytes are zeroed out. https://github.com/FuelLabs/sway/blob/master/sway-lib-std/src/vm/evm/evm_address.sw#L21

So I think using __addr_of here will point to the zeroes.
Maybe we need to add `write_evm_address` / `read_evm_address` functions to `bytes-extended`, and similarly implement for `EvmAddress` the `packed_bytes` / `from_packed_bytes` fns that account for this 12 byte left padding so we point to the 20 bytes with the content

Reply:
left padding sucks